### PR TITLE
Move control of qualification assignment to Experiment class

### PR DESCRIPTION
**Pull request by jessesnyder**
## Description
Addresses issue #2261 

Allow the Experiment to control qualification assignment to workers.

Instead of the Dallinger infrastructure calling the recruiter directly for qualification assignment, this work will now manged by the Experiment in cooperation with the recruiter for a Participant.

## Motivation and Context
The underlying objective is to add flexibility to when MTurk Qualifications are assigned to workers. Currently this is always done after the post-experiment questionnaire is submitted to the `/worker_completed` route. 

The PR proposes moving in the direction of notifying the Experiment class when various experiment lifecycle events happen, and letting the Experiment then request qualification assignment from participant's recruiter reference (to support the MultiRecruiter).

### Backwards compatibility

The status quo is preserved by:
1. adding a new method to the Experiment class (called `participant_task_completed()`) which will be called _instead of_ `Recruiter.notify_completed()` from the `/worker_completed` route, and
2. adding a new public method `assign_experiment_qualifications()` to the `Recruiter` interface.

The default implementation of `Experiment.participant_task_completed()` will just call `Recruiter.assign_experiment_qualifications()` with the same qualifications the recruiter currently manages internally.

Responsibility for parsing config.txt recruitment-related values (`assign_qualifications`, `group_name`) now lies with the Experiment, via a set of new properties: `qualifications_active`, `qualifications`, and `group_qualifications`. The default implementation of these properties preserves the current default behavior by parsing config.txt and translating these into qualification request definitions compatible with the Recruiter's API. However, these can be overridden to provide values in a different way (most obviously, dispensing with config.txt altogether and just defining these values in python.)

### Early (or Multiple) Qualification Assignment

The Experiment class is already called at multiple points, starting from the very beginning of the Participant lifecycle:
1. `create_participant(self, worker_id, hit_id, assignment_id, mode, recruiter_name=None, fingerprint_hash=None)`
2. `create_node(self, participant, network)`
3. `node_post_request(self, participant, node)`

At any of these points, the experiment could assign one or more of the qualifications to the participant's worker ID.

There is also at least one other opportunity to assign qualifications late, without adding any new methods to Experiment:
1. `submission_successful(self, participant)`

### Qualification Creation
We use the `group_name` config variable for defining custom qualifications, and currently assume it's a single value This PR adds support for a comma-separated list.

Further, this PR changes qualification creation to be just-in-time, so that experiments can determine qualifications at run-time.

## Other fixes
* `/worker_complete` route is now idempotent (no-op if `participant.end_time` is not `None`)
* form freezing/disabling fixed for questionnaire

## How Has This Been Tested?
[x] Automated tests
[x] MTurk sandbox with Bartlett1932

## Screenshots (if appropriate):
Bartlett1932 tested over a few runs, with the following config:
```
...
assign_qualifications = True
group_name = red1,green1
...
```

and the following method override:
```python
    def participant_task_completed(self, participant):
        """[blah blah blah]
        """
        if not self.qualification_active:
            logger.info("Qualification assignment is globally disabled; ignoring.")
            return

        if participant.status == "overrecruited":
            return

        quals = self.qualifications
        quals.append({"name": "another", "description": "has a score", "score": 21})
        participant.recruiter.assign_experiment_qualifications(
            worker_id=participant.worker_id, qualifications=quals
        )
```

I was assigned the expected qualifications:

Assigned_Qualifications_-_Amazon_Mechanical_Turk

### CODE REVIEW

File: dallinger/recruiters.py [hunk 1]
```diff
@@ -561,6 +555,20 @@ def open_recruitment(self, n=1):
             "message": "HIT now published to Amazon Mechanical Turk",
         }
 
+    def assign_experiment_qualifications(self, worker_id, qualifications):
+        """Assigns MTurk Qualifications to a worker. Do nothing if qualification
+        assignment is globally disabled.
+        """
+        if not self.qualification_active:
```
Comment (jessesnyder) [inline dallinger/recruiters.py:562]: I might be inclined to move this check to the experiment also. Let the recruiter be ignorant about all the policies about if and when to assign qualifications.

File: dallinger/experiment.py [hunk 2]
```diff
@@ -408,6 +421,25 @@ def attention_check(self, participant):
...
+
+        Assign the qualifications to the Participant, via the recruiter.
+        These will include one Qualification for the experiment
+        ID, and others for the configured group_name, if it's been set.
+
+        Overrecruited participants don't receive qualifications, since they
+        haven't actually completed the experiment. This allows them to remain
+        eligible for future runs. Blah blah blah.. just an example.
+        """
+        if participant.status == "overrecruited":
```
Comment (thomasmorgan) [inline dallinger/experiment.py:436]: Is "overrecruited" a base status? I didn't know it existed which makes me a little queasy to see other code relying on it. I assume this is a way to not give participants the qualification if they didn't meaningfully take part. I wonder if there is a way to do this without relying on the status code, which users might not use correctly (I wouldn't have in this case).

File: dallinger/recruiters.py (hunk 1 above)
Comment (thomasmorgan) [reply dallinger/recruiters.py:562]: Agreed. The recruiter is just a middle man, the experiment should make the decisions.

**Comment by thomasmorgan** [comment]
Hey Jesse, this looks good.

Why do we need to specify the names of potential qualifications up front in config.txt? Is this so they can be pre-made at deployment? Why not bake this functionality into the qualification assignment function itself? i.e. the qualification function first checks the qualification exists, if it doesn't it makes it, and then either way it assigns it. This way users can assign any qualification at any time, even permitting qualifications whose name cannot be easily determined in advance (although I can't think of why this might be right now).

File: dallinger/experiment.py (hunk 2 above)
Comment (jessesnyder) [reply dallinger/experiment.py:436]: It is. See https://github.com/Dallinger/Dallinger/blob/master/dallinger/models.py#L238 It was introduced once Jordan and others found significant overrecruitment to be an effective strategy for getting a quorum in large waiting room based experiments.

**Comment by jessesnyder** [comment]
> Hey Jesse, this looks good.
> 
> Why do we need to specify the names of potential qualifications up front in config.txt? Is this so they can be pre-made at deployment? Why not bake this functionality into the qualification assignment function itself? i.e. the qualification function first checks the qualification exists, if it doesn't it makes it, and then either way it assigns it. This way users can assign any qualification at any time, even permitting qualifications whose name cannot be easily determined in advance (although I can't think of why this might be right now).

Ah, thanks for pointing this out. I had it in my head that the qualification IDs are (and need to be) included in the HIT definition in some way, but this isn't the case (I think I was conflating _required_ qualifications, which obviously do). So, I see no reason not to create them "just in time", and I agree this would be better, given the direction we're going.

File: dallinger/experiment.py (hunk 2 above)
Comment (jessesnyder) [reply dallinger/experiment.py:436]: I just noticed it's missing from the comment block that generates the Sphinx documentation. I don't know if this was intentional. :-/

**Comment by pmcharrison** [comment]
Nice job, @thomasmorgan and @jessesnyder - excited to see this progress!

File: dallinger/experiment.py (hunk 2 above)
Comment (thomasmorgan) [reply dallinger/experiment.py:436]: Got it. I guess it's kinda my fault. I've never used the built in waiting room and just make my own, so I've never bumped into any of this code. It would mean that this check would fail for me though, but that at least wouldn't break my intuition for qualification handling, I would just miss out on a few things.

**Comment by thomasmorgan** [comment]
OK, great. This is another thing where the relationship between config.txt and experiment.py might confuse users: it's not necessarily clear why the list of required qualifications should go in config.txt, especially given that assigning qualifications is handled in experiment.py. This does become clear once you realize that the HIT is configured before the experiment is loaded, but its something we should bear in mind for documentation.

**Review by alecpm — CHANGES_REQUESTED** [review]
I'm pretty sure this needs quite a few test changes, but otherwise it looks good to me.

**Comment by jessesnyder** [comment]
@pmcharrison @jacobyn This is ready to try out.

**Comment by pmcharrison** [comment]
> @pmcharrison @jacobyn This is ready to try out. PR description also updated.

Thanks, @jessesnyder. We're going to work on it this month and can get back to you for the next sprint.

**Comment by pmcharrison** [comment]
Thanks again, @jessesnyder! @jacobyn and I discussed for a long time, and we're happy but would like to request a couple of things:

1. Can we extend `Recruiter.assign_experiment_qualifications` so that it contains an optional list of values to assign?
2. We discussed and actually it does complicate things in many ways to force the qualifications to be created in advance. Can we *try* putting some kind of wait process in `_ensure_mturk_qualifications` so that it ensures that the qualification is created *and* available before proceeding to assign the qualification? This would simplify things so much! Hopefully it would only be a matter of 500 ms or so.

**Comment by jessesnyder** [comment]
@pmcharrison This now supports assigning specific qualification values, and retries to make sure qualifications are available for assignment on MTurk. To test, I used Bartlett1932 over a few runs, with the following config:
```
...
assign_qualifications = True
group_name = red1,green1
...
```

and the following method override:
```python
    def participant_task_completed(self, participant):
        """[blah blah blah]
        """
        if not self.qualification_active:
            logger.info("Qualification assignment is globally disabled; ignoring.")
            return

        if participant.status == "overrecruited":
            return

        quals = self.qualifications
        quals.append({"name": "another", "description": "has a score", "score": 21})
        participant.recruiter.assign_experiment_qualifications(
            worker_id=participant.worker_id, qualifications=quals
        )
```

I was assigned the expected qualifications:

Assigned_Qualifications_-_Amazon_Mechanical_Turk

**Comment by pmcharrison** [comment]
It looks great, @jessesnyder , thank you! Did you get a feel for how many retries were necessary for a new qualification?

**Comment by jessesnyder** [comment]
@pmcharrison I never needed a retry. Just querying may introduce enough of a delay?

File: dallinger/experiment_server/experiment_server.py [hunk 3]
```diff
@@ -1556,18 +1556,24 @@ def worker_complete():
...
 
 def _worker_complete(participant_id):
-    participants = models.Participant.query.filter_by(id=participant_id).all()
-    if not participants:
+    participant = models.Participant.query.get(participant_id)
+    if not participant:
         raise KeyError()
 
-    participant = participants[0]
+    if participant.end_time is not None:  # Provide idempotence
```
Comment (alecpm) [inline dallinger/experiment_server/experiment_server.py:1563]: I think this is safe because this `_worker_complete` function seems to be called before the `worker_function` which calls various `WorkerEvent` instances (like `AssignmentSubmitted`) which also set `end_time` via the `update_particant_end_time` method (note typo). However, I'm a little wary that we're still setting/updating this in multiple places.

File: dallinger/frontend/static/scripts/dallinger2.js [hunk 4]
```diff
@@ -632,9 +633,10 @@ var dallinger = (function () {
       number: 1,
       response: JSON.stringify(formDict)
     });
-    spinner.freeze([$('form :input')]);
-    xhr.done(dlgr.submitAssignment);
-    xhr.always(function () { spinner.unfreeze(); });
+    spinner.freeze([$inputs, $button]);
+    xhr.done(function () {
+      dlgr.submitAssignment().always(function () { spinner.unfreeze(); });
```
Comment (alecpm) [inline dallinger/frontend/static/scripts/dallinger2.js:638]: Won't this prevent `unfreeze` if the POST succeeds, but the `submitAssignment` fails? Is that desirable?

File: dallinger/mturk.py [hunk 5]
```diff
@@ -374,7 +405,7 @@ def revoke_qualification(self, qualification_id, worker_id, reason=""):
             )
         )
 
-    def get_qualification_score(self, qualification_id, worker_id):
+    def current_qualification_score(self, qualification_id, worker_id):
```
Comment (alecpm) [inline dallinger/mturk.py:408]: I guess this is as good a time as any to rename things, but I'm not sure I follow the distinction being made here.

File: dallinger/recruiters.py [hunk 6]
```diff
@@ -615,6 +607,30 @@ def open_recruitment(self, n=1):
...
+    def assign_experiment_qualifications(self, worker_id, qualifications):
+        """Assigns MTurk Qualifications to a worker.
+
+        @param worker_id       string  the MTurk worker ID
+        @param qualifications  list of dict w/   `name`, `description` and
+                               (optional) `score` keys
+        """
+        by_name = {qual["name"]: qual for qual in qualifications}
+        result = self._ensure_mturk_qualifications(qualifications)
+        for qual in result["new_qualifications"]:
```
Comment (alecpm) [inline dallinger/recruiters.py:619]: Shouldn't this check `available` for these qualifications, since some may not have been available after 5 tries?

File: tests/test_experiment.py [hunk 7]
```diff
@@ -15,7 +17,8 @@ def klass(self):
         return Experiment
 
     @pytest.fixture
-    def exp(self, klass):
+    def exp(self, active_config, klass):
```
Comment (alecpm) [inline tests/test_experiment.py:20]: Is this change needed?

**Review by alecpm — COMMENTED** [review]
I have a few questions and concerns, but mostly this looks good.

File: dallinger/experiment.py [hunk 8]
```diff
@@ -187,6 +187,67 @@ def recruiter(self):
...
+        experiment_qualification_desc = "Experiment-specific qualification"
+        config = get_config()
+        quals = [
+            {
+                "name": config.get("id"),
+                "description": experiment_qualification_desc,
+            }
+        ]
+
+        return quals + self.group_qualifications
```
[the remark this answers is not in the document]
Comment (jessesnyder) [reply dallinger/experiment.py:214]: My rationale was that you're unlikely to want to override the experiment ID-based qualification, and are more likely to override the group qualifications (mostly obviously, not using the `group_name` config variable, but just defining the qualifications in the property instead), but I agree it's gotten complicated and may not be worth it.

File: dallinger/recruiters.py [hunk 9]
```diff
@@ -127,12 +132,6 @@ def reward_bonus(self, assignment_id, amount, reason):
         """Throw an error."""
         raise NotImplementedError
 
-    def notify_completed(self, participant):
```
[the remark this answers is not in the document]
Comment (jessesnyder) [reply dallinger/recruiters.py:130]: It wasn't really public, as it was only called from Dallinger.  I share your concerns about the new method being less generic. I think there might be a couple mitigating factors:
1. The relationships have been inverted, and the "generic" call is now from `experiment_server` to the experiment's `participant_task_completed(participant)` method. This allows the experiment to tailor the instructions to the particular recruiter it's using, which seems inevitable, and more realistic than trying to provide a truly standard interface for all recruiters (if you knew your recruiter provided some interface unique to it, you could still call it from your experiment).
2. I've forgotten the second one!

File: dallinger/recruiters.py (hunk 6 above)
Comment (jessesnyder) [reply dallinger/recruiters.py:619]: I meant to come back to this issue and forgot. I can't decide if it should check here, or do something (log an error, I guess?) at the end of `_ensure_mturk_qualifications()`.  Went with the latter.

File: dallinger/experiment_server/experiment_server.py (hunk 3 above)
Comment (jessesnyder) [reply dallinger/experiment_server/experiment_server.py:1563]: Yeah, this is a mess because (perhaps among other reasons) MTurk participants are a special case for which we defer all the `WorkerEvent` tasks until word comes in from MTurk that the HIT has been submitted. I'm more worried about this the more I think about it (some experiments calculate bonuses based in part on `end_time`) but I'm not sure it makes a difference for this PR.

File: dallinger/frontend/static/scripts/dallinger2.js (hunk 4 above)
Comment (jessesnyder) [reply dallinger/frontend/static/scripts/dallinger2.js:638]: I don't think so, because the unfreeze is in an `always()` call on the return of `submitAssignment()`, but it _will_ prevent unfreeze if the first POST fails.  The problem is we don't want to unfreeze before the `submitAssignment` call finishes.

Comment (alecpm) [reply dallinger/frontend/static/scripts/dallinger2.js:638]: Yeah, the POST fails so `submitAssignment()` call never happens is the case I was thinking about when I decided to comment and then I wrote the opposite for some reason. So this is a change in behavior. I guess if the POST fails we expect the normal error handling process and don't worry about `unfreeze`?

File: dallinger/experiment.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alecpm) [reply dallinger/experiment.py:214]: Right that makes sense, but if both methods have default implementations that would need to be replicated in a sub-class it seems like we're making things more complicated than they need to be. It seems like a `qualifications()` method would suffice, especially since most use cases would just add some additional qualifications to the list pulled in with a `super()` call. A property seems like an unnecessary complication that makes documenting it more difficult. Is there a use case where this wouldn't be dynamic and someone could just set the attribute on the class?

File: docs/source/the_experiment_class.rst [hunk 10]
```diff
@@ -30,6 +30,15 @@ what to do with the database when the server receives requests from outside.
   .. autoattribute:: recruiter
     :annotation:
 
+  .. autoattribute:: qualifications
+    :annotation:
+
+  .. autoattribute:: group_qualifications
+    :annotation:
+
+  .. autoattribute:: qualification_active
```
[the remark this answers is not in the document]
Comment (alecpm) [reply docs/source/the_experiment_class.rst]: Yeah, I get that most experimenters probably don't want to write semi-generic parametrizable experiments, so the config stuff is a complication. At this point the alternative is having a bunch of different ways to do the same thing, and I'm not sure that's an improvement. This attribute is only checked in one place and it's not currently public API, so I would prefer getting rid of it entirely.

File: dallinger/recruiters.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alecpm) [reply dallinger/recruiters.py:130]: The change looks good to me and seems like a better approach. I'm a just little worried that people might be relying on this in some way, but I guess almost no one has custom recruiters, so it's probably not a big deal.

File: dallinger/experiment_server/experiment_server.py (hunk 3 above)
Comment (alecpm) [reply dallinger/experiment_server/experiment_server.py:1563]: Yeah, this seems like it should always be OK since it's always fired before any of those `WorkerEvent` things get called. The whole things seems a little brittle to me, especially if we're using end_time for anything important like bonus payments.

**Review by alecpm — APPROVED** [review]
With the unavailable qualification fix this looks good to go, so I Approve. Let me know if you'd like to discuss the `qualifications` property stuff though.

File: dallinger/frontend/static/scripts/dallinger2.js (hunk 4 above)
Comment (jessesnyder) [reply dallinger/frontend/static/scripts/dallinger2.js]: This is thoroughly tested and working correctly now (revisions made).

Comment (jessesnyder) [reply dallinger/frontend/static/scripts/dallinger2.js]: I couldn't stand looking at the crappy error forms any more so tidied those up a little bit also.

File: dallinger/mturk.py (hunk 5 above)
Comment (jessesnyder) [reply dallinger/mturk.py:408]: Seemed critical at the time for some reason, but in retrospect it doesn't seem to make much of a difference.

File: tests/test_experiment.py (hunk 7 above)
Comment (jessesnyder) [reply tests/test_experiment.py:20]: No, forgot to remove that after an experiment didn't pan out. Removed.

File: dallinger/experiment.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jessesnyder) [reply dallinger/experiment.py:214]: > Is there a use case where this wouldn't be dynamic and someone could just set the attribute on the class?
That's what I had in mind, based on a vague recollection of someone wanting to do that (set class attributes for this kind of thing).

Still, I think we need to fight back complication where possible, so I'm inclined to reduce this all to a single method as you suggest. I'll ask the group tomorrow. <reminder: @pmcharrison @jacobyn @skleinfeldt >

File: dallinger/experiment_server/experiment_server.py (hunk 3 above)
Comment (jessesnyder) [reply dallinger/experiment_server/experiment_server.py:1563]: I'm not sure why the single `end_time` setting at the beginning of `_worker_complete()` couldn't be the one and only place this happens, but we'll probably need to do give it a thorough review.

File: docs/source/the_experiment_class.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (jessesnyder) [reply docs/source/the_experiment_class.rst]: Persuaded.
